### PR TITLE
Tensor View, Vectorise refactor and CBOW example

### PR DESCRIPTION
**Pull request by troels-im**
This PR contains a number of changes:

- Refactoring vectorise
- Implementing Tensor view
- Moving CBOW across using graph to define computation.

### CODE REVIEW

**Review by markos — DISMISSED** [review]
This is a huge PR and I cannot comment on CBOW example, but the rest, in particular the vectorise changes, look good to me.

File: libs/vectorise/include/vectorise/memory/parallel_dispatcher.hpp [hunk 1]
```diff
@@ -629,6 +629,7 @@ class ParallelDispatcher : public ConstParallelDispatcher<T>
     }
   }
 
+  // TODO: This function is potentially slow
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/memory/parallel_dispatcher.hpp]: TODO's need issue numbers

File: libs/vectorise/include/vectorise/memory/parallel_dispatcher.hpp [hunk 2]
```diff
@@ -769,13 +766,14 @@ class ParallelDispatcher : public ConstParallelDispatcher<T>
     }
   }
 
+  // TODO: Remove below
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/memory/parallel_dispatcher.hpp]: issue number

File: libs/vectorise/include/vectorise/arch/sse/register_float.hpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+FETCH_ADD_OPERATOR(==, float, __m128, _mm_cmpeq_ps)
+FETCH_ADD_OPERATOR(!=, float, __m128, _mm_cmpneq_ps)
+FETCH_ADD_OPERATOR(>=, float, __m128, _mm_cmpge_ps)
+FETCH_ADD_OPERATOR(>, float, __m128, _mm_cmpgt_ps)
+FETCH_ADD_OPERATOR(<=, float, __m128, _mm_cmple_ps)
+FETCH_ADD_OPERATOR(<, float, __m128, _mm_cmplt_ps)
+
+#undef FETCH_ADD_OPERATOR
+
+// Manage NaN
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/arch/sse/register_float.hpp]: is this meant to be a TODO with an issue number

File: libs/vectorise/include/vectorise/arch/sse/register_float.hpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+inline float reduce(VectorRegister<float, 128> const &x)
+{
+  __m128 r = _mm_hadd_ps(x.data(), _mm_setzero_ps());
+  r        = _mm_hadd_ps(r, _mm_setzero_ps());
+  return _mm_cvtss_f32(r);
+}
+
+/*
+// TODO: Float equivalent
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/arch/sse/register_float.hpp]: issue number

File: libs/vectorise/include/vectorise/arch/sse/register_double.hpp [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+inline double first_element(VectorRegister<double, 128> const &x)
+{
+  return _mm_cvtsd_f64(x.data());
+}
+
+// Floats
+
+// TODO(unknown): Rename and move
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/arch/sse/register_double.hpp]: TODO(unknown) presumably means you need to open a new issue?

File: libs/vectorise/include/vectorise/arch/sse/register_double.hpp [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+  __m128d          imm  = _mm_cmplt_pd(a.data(), b.data());
+  __m128i          ival = _mm_castpd_si128(imm);
+  constexpr double done = double(1);
+  const __m128i    one  = _mm_castpd_si128(_mm_load_pd1(&done));
+  __m128i          ret  = _mm_and_si128(ival, one);
+  return VectorRegister<double, 128>(_mm_castsi128_pd(ret));
+}
+
+// Manage NaN
+//__m128d _mm_cmpord_pd (__m128d a, __m128d b)
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/arch/sse/register_double.hpp:215]: commented out code here looks like its referring to a TODO + issue number?

File: libs/vectorise/include/vectorise/arch/sse/math/approx_reciprocal.hpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+inline VectorRegister<float, 128> approx_reciprocal(VectorRegister<float, 128> const &x)
+{
+
+  return VectorRegister<float, 128>(_mm_rcp_ps(x.data()));
+}
+
+inline VectorRegister<double, 128> approx_reciprocal(VectorRegister<double, 128> const &x)
+{
+  // TODO(issue 3): Test this function
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/arch/sse/math/approx_reciprocal.hpp]: i think this issue number is probably wrong. looking at our issues #3 is closed already and reads: "Separate io_service into common class with threading #3"

File: libs/ml/examples/cbow_simple/word_loader.hpp [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+      ret.first(i, 0) = -1;
+    }
+
+    // Creating negatice sample
+    for (uint64_t i = 1; i < negative_samples_; ++i)
+    {
+      ret.second(i, 0) = static_cast<T>(SampleNegative(center_word));
+    }
+
+    // TODO: Divide into sentences
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/word_loader.hpp]: i recognise this isn't your original work, so you probably didn't put this TODO here. but if we're merging into the main library this work then it should be ready to go. so this TODO is either to be ditched or have an issue number assigned - probably just ditch it

File: libs/ml/examples/cbow_simple/word_loader.hpp [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+      {
+        removed_words.push_back(word);
+      }
+      corpus_[j] = std::move(word);
+    }
+
+    corpus_.resize(j);
+
+    // Reindexing
+    /*
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/word_loader.hpp]: commented out code for removal?

File: libs/ml/examples/cbow_simple/word_loader.hpp [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/byte_array/byte_array.hpp"
+#include "crypto/fnv.hpp"
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/word_loader.hpp]: what's this include for?

File: libs/ml/examples/cbow_simple/word_loader.hpp [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+#include "math/tensor.hpp"
+
+#include <set>
+#include <unordered_map>
+#include <vector>
+namespace fetch {
+namespace ml {
+
+template <typename T>
+class WordLoader : public DataLoader<fetch::math::Tensor<T>, fetch::math::Tensor<T>>
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/word_loader.hpp]: no function comments anywhere makes it pretty un-fun to read if you don't already know how this works

File: libs/ml/examples/cbow_simple/main.cpp [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+#include "unigram_table.hpp"
+#include "w2v_cbow_dataloader.hpp"
+#include "word_loader.hpp"
+
+using namespace std::chrono;
+using namespace fetch::ml;
+using FloatType = float;
+
+CBOWLoader<FloatType> data_loader(5, 25);
+WordLoader<FloatType> new_loader;
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/main.cpp]: please correct me if i'm wrong but it looks like this WordLoader is only instantiated here and never used; and then the whole WordLoader class is never used anywhere else?

File: libs/ml/examples/cbow_simple/main.cpp [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+double time_step     = 0;
+
+uint64_t                          dimensionality = 200;  // 200
+uint64_t                          iter           = 1;
+FloatType                         alpha          = static_cast<FloatType>(0.025);
+FloatType                         starting_alpha;
+int32_t                           negative = 25;
+high_resolution_clock::time_point last_time;
+
+std::string readFile(std::string const &path)
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/main.cpp]: ReadFile

File: libs/ml/examples/cbow_simple/main.cpp [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+{
+  std::ifstream t(path);
+  return std::string((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+}
+
+void PrintStats(uint32_t const &i, uint32_t const &iterations)
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/main.cpp]: this does more than just print stats! it updates the learning rate in a linear step decay

File: libs/ml/examples/cbow_simple/main.cpp [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+  {
+    return 0;
+  }
+
+  output_file = "./vector.bin";
+  train_file  = argv[1];
+
+  alpha = static_cast<FloatType>(0.05);  // Initial learning rate
+
+  //  std::cout << "New loader" << std::endl;
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/main.cpp]: leftover commented out code

File: libs/ml/examples/cbow_simple/main.cpp [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+{
+  if (argc == 1)
+  {
+    return 0;
+  }
+
+  output_file = "./vector.bin";
+  train_file  = argv[1];
+
+  alpha = static_cast<FloatType>(0.05);  // Initial learning rate
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/main.cpp]: you have a variable called starting_alpha but you also hardcode it here?

File: libs/ml/examples/cbow_simple/main.cpp [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+
+  //  std::cout << "New loader" << std::endl;
+  //  new_loader.Load(train_file);
+
+  std::cout << "Old loader" << std::endl;
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/main.cpp]: leftover cout?

File: libs/ml/examples/cbow_simple/polyfill.hpp [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+#include "math/tensor.hpp"
+#include "meta/is_iterable.hpp"
+
+namespace fetch {
+namespace math {
+
+template <typename T1, typename T2>
+fetch::meta::IsIterableTwoArg<T1, T2, void> PolyfillInlineAdd(T1 &ret, T2 const &other)
+{
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/polyfill.hpp:28]: this looks like vectorised code to add a tensor slice by slice to another tensor. Would be nice to have a comment that says this is what it does. I may be wrong but i'm not aware of polyfill being a standard name for this operation (I don't see it in numpy and wiki's entry on polyfill(programming) is unrelated), so I would say a comment is worthwhile

File: libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+    std::vector<std::string> words;
+    for (std::stringstream ss(result); ss >> word;)
+    {
+      words.push_back(word);
+    }
+    return words;
+  }
+
+private:
+  uint64_t                                             currentSentence_;
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp]: current_sentence_

File: libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+  uint64_t                                             currentWord_;
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp]: current_word_

File: libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+      {
+        auto value = vocab_.insert(std::make_pair(s, std::make_pair((uint64_t)(vocab_.size()), 0)));
+        indexes.push_back((*value.first).second.first);
+        value.first->second.second++;
+      }
+    }
+    return indexes;
+  }
+
+  std::vector<std::string> PreprocessString(std::string const &s)
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp:217]: it may be that for text8 it doesn't make any difference, but strictly the google code preprocesses by removing tabs, spaces, and return carriages.

File: libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+        return kvp.first;
+      }
+    }
+    return "";
+  }
+
+private:
+  std::vector<uint64_t> StringsToIndexes(std::vector<std::string> const &strings)
+  {
+    std::vector<uint64_t> indexes;
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp:203]: pedantic, i know, but it's indices (indexes is not english, well, maybe it's a verb but not a noun)

File: libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+//
+//------------------------------------------------------------------------------
+
+#include "core/random/lcg.hpp"
+#include "dataloader.hpp"
+#include "math/tensor.hpp"
+#include "unigram_table.hpp"
+
+#include <exception>
+#include <fstream>
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp]: you don't do any file loading here i think

File: libs/ml/examples/cbow_simple/dataloader.hpp [hunk 24]
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
Comment (baykaner) [inline libs/ml/examples/cbow_simple/dataloader.hpp]: I think this is an exact copy of our abstract definition of a dataloader: ml/include/ml/dataloaders/dataloader.hpp
if so, please delete this and inherit from that one

File: libs/math/include/math/tensor_declaration.hpp [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+#include "vectorise/memory/array.hpp"
+#include "vectorise/memory/shared_array.hpp"
+
+namespace fetch {
+namespace math {
+
+template <typename DataType, typename ContainerType = memory::SharedArray<DataType>>
+class Tensor;
+
+// TODO: Migrate to:
```
Comment (baykaner) [inline libs/math/include/math/tensor_declaration.hpp]: issue number

File: libs/math/include/math/tensor.hpp [hunk 26]
```diff
@@ -78,9 +79,9 @@ class Tensor
 
   static constexpr char const *LOGGING_NAME = "Tensor";
 
-  enum
+  enum  // TODO: Get from view
```
Comment (baykaner) [inline libs/math/include/math/tensor.hpp]: issue number

File: libs/math/include/math/tensor.hpp [hunk 27]
```diff
@@ -295,10 +296,8 @@ class Tensor
   template <typename OtherType>
   Tensor operator/=(OtherType const &other);
 
-  Tensor &DotTranspose(Tensor const &A, Tensor const &B, Type alpha = 1.0, Type beta = 0.0);
-  Tensor &TransposeDot(Tensor const &A, Tensor const &B, Type alpha = 1.0, Type beta = 0.0);
-  Type    Sum() const;
-
+  // TODO: Make free functions
```
Comment (baykaner) [inline libs/math/include/math/tensor.hpp]: issue number

File: libs/math/include/math/ml/loss_functions/cross_entropy.hpp [hunk 28]
```diff
@@ -59,17 +59,16 @@ typename ArrayType::Type CrossEntropyLoss(
...
     auto     x_it = x.cbegin();
     auto     y_it = y.cbegin();
     DataType one{1};
-    DataType zero{0};
     DataType tmp;
 
     while (x_it.is_valid())
     {
-      ASSERT((*y_it == one) || (*y_it == zero));
+      assert((*y_it == one) || (*y_it == static_cast<DataType>(0)));
```
Comment (baykaner) [inline libs/math/include/math/ml/loss_functions/cross_entropy.hpp:64]: this variable declaration of "zero" and "one" was because it seemed needless to continually instantiate/cast a variable in a while loop. but presumably you're changed it back to a cast because the compiler can be relied upon to optimise this out?

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/math/include/math/ml/loss_functions/cross_entropy.hpp (hunk 28 above)
Comment (troels-im) [reply libs/math/include/math/ml/loss_functions/cross_entropy.hpp:64]: It leaves the variable unused in release mode as `zero` was only used here. It definitely has no performance implications for release mode and most likely not for debug mode.

File: libs/ml/examples/cbow_simple/word_loader.hpp (hunk 10 above)
Comment (troels-im) [reply libs/ml/examples/cbow_simple/word_loader.hpp]: hashing of `ByteArray`

**Review by baykaner — APPROVED** [review]
